### ISSUE TITLE
Feat - voice manager mode for the notestepper

### ISSUE THREAD

**Issue by PowerUser64** [BespokeSynth/BespokeSynth#1289]
## Summary

Currently, the only way to get polyphony from bespoke is to duplicate the thing you want to have multiple voices so there's one voice for each note, and use something like the notestepper to distribute notes between them. The problem with this is it's not very smart. If you're holding down a low note to play high notes over it, you'll only be able to play as many notes as you have voices, at which point the notestepper will steal the voice from the low note.

To solve this, an option called "voice manager mode" or similar should be added as a toggle to the notestepper's triangle menu. In this mode, the notestepper would keep track of which destinations don't have notes being sent to them, and when picking the next cable to send a note to, it would send it to the first unused cable, or the oldest currently-used cable, so held notes will only have their voices stolen if the number of held notes exceeds the number of voices.

Perhaps other voice management modes could be added in the future, but I think this mode is the most common form of voice management.

### Motivation

This would make it easier to create synthesizers with bespoke and convert them into polyphonic synthesizers.

**Comment by PowerUser64** [comment]
Page 83 of the [Arturia MiniFreak manual](https://dl.arturia.net/products/minifreak/manual/minifreak_Manual_1_0_0_EN.pdf) has a good section that outlines a few voice allocation types:

![image](https://github.com/BespokeSynth/BespokeSynth/assets/24578572/b776f225-d9ab-40a3-b99f-f5349c3d96d4)

The "Cycle" mode is more or less what is outlined above. I own a MiniFreak, and I have found that the "Reassign" mode can be very handy as well, and it accomplishes a very similar thing to the "Reset" mode.

I think the reset mode might be the easiest to implement.

**Comment by embix** [comment]
The low hanging fruit would be: if there still are free voices, don't steal. If there aren't: ignore. That alone would be useful (for me at least)

Next deliverable would be: If there aren't, steal the oldest. That should be useful to other people as well, maybe even what they'd expect.

Then come your cited polyphony modes. After that unison-n and mono-m modes come to mind (maybe even with intentional detuning via some control input).

TL;DR: my suggestion is to strip this feature into smaller deliverables that already add value without being overwhelming to implement all at once.

**Comment by NoxyNixie** [comment]
I think I'd rather make the voice manager a new module, that way you can have the controls in module instead of all hidden away in the triangle menu.

**Comment by PowerUser64** [comment]
I agree. At the time, I thought the notestepper would be a good place to add this because it has semi-overlapping features, but in hindsight, it makes more sense for it to be separate.
